README.md
# Rotate

[![Golang](https://img.shields.io/badge/Golang-1.18-blue)](https://go.dev/)
[![GitHub](https://img.shields.io/github/license/huashusu/rotate)](https://github.com/Huashusu/rotate)

[English](README.md) | [简体中文](README.zh_CN.md)

## Preconditions

- **[Go](https://go.dev/)** version >= 1.18

> If you have tested it and can run it under a lower version, you can raise an `issue` or `pull requests` to change the
> docs and go.mod.

## Getting rotate

```shell
go get -u github.com/huashusu/rotate
```

## Documentation

[Document](https://godoc.org/github.com/Huashusu/rotate)

## Change log

[Change log](CHANGELOG.md)

## Example

```go
package main

import (
	"log"
	"os"
	"github.com/huashusu/rotate"
	"time"
)

func main() {
	file, err := rotate.New("log", "2006-01-02/15-04-05", "log",
		rotate.WithRotationDuration(time.Minute*1),
		rotate.WithDeleteEmptyFile(true),
		rotate.WithMaxSize(rotate.MB*10),
		rotate.WithMaxAge(time.Minute*5),
		rotate.WithExpiredHandler(func(expFiles []string) {
			for i := 0; i < len(expFiles); i++ {
				os.Remove(expFiles[i])
			}
		}),
	)
	if err != nil {
		panic(err)
	}
	log.SetOutput(file)
}
```

`New` method signature: `func New(dir, layout, ext string, options ...Option) (*Rotate, error)`

Parameter Description:

- `dir`: The root directory where log files are stored.
- `layout`: For the time template used in the generated log, the syntax used is the `format` method of the
  built-in `time` package, and there should be no illegal characters in the file name. The example is as follows:
    - No directory hierarchy exists: layout: `2006-01-02-15-04-05`
    - Separate logs by month: layout: `2006-01-02/15-04-05`
    - Sort logs by year, month and day: layout: `2006/01/02/15-04-05`
- `ext`: Log file suffix

## Option Parameters

### Set split time(`WithRotationDuration`)`default: Day`

**explain:** A truncating design was used for time selection. If the setting is less than one day, it is best to be
divisible by 24 (hours). The case is as follows:

current time: `2023/09/12 15:45:40`

- Set for 10 minutes: `2023/09/12 15:40:00`
- Set for 30 minutes: `2023/09/12 15:30:00`
- Set 1 hour: `2023/09/12 15:00:00`
- Set 6 hours: `2023/09/12 12:00:00`
- Set 1 day: `2023/09/12 00:00:00`

### Set time zone(`WithTimeZone`)`default: time.Local`

**explain:** The time template does not need to carry the time zone flag, and is set separately through the time zone,
which is used to truncate the time and parse the time from the file name.

### Set expiration time(`WithMaxAge`)`default: 0`

**explain:** Every time a file is split, it will be checked to see if there are any expired files.

### Set file size limit(`WithMaxSize`)`default: 0`

**explain:** After writing exceeds the file size, it will be split into log files with serial numbers.

### Set expiration handler(`WithExpiredHandler`)`default: nil`

**explain:** Usually used together with the expiration time, the path of the expired file is passed in, for example:
compressing the expired file, reading the file and sending it, deleting the file, etc.

### Set delete empty files(`WithDeleteEmptyFile`)`default: true`

**explain:** After the log is split, if the last file size is 0, delete it.

### Set delete empty directory(`WithDeleteEmptyDir`)`default: true`

**explain:** After deleting empty file logs, there may be empty folders. You can use this method to clean up empty
folders.

## Const

### Used for file size limits

`KB`: `1024`

`MB`: `1048576`

`GB`: `1073741824`

### Used for expiration time settings

`Day`: `time.Hour * 24`

`Week`: `Day * 7`

`Month`: `Day * 30`

## Static Method

`SetSymbol`: Set the left and right wrapping symbols of the serial number in the file, default
value:`Left`=`[` `Right`=`]`

`SetPerm`: Set the permission mode for creating folders and files, default value: `0644`

CHANGELOG.md
# Change log

## v0.0.2

- [NEW] Add option to delete empty folders
- [FIX] Modify the dependent username of the library to lowercase

## v0.0.1

- [NEW] Rotate structure
- [NEW] Set file serial number wrapping symbol`SetSymbol`
- [NEW] Set the permission mode for creating folders and files`SetPerm`
- [NEW] Built in common file size constants`KB` `MB` `GB`
- [NEW] Built in commonly used long intervals`Day` `Week` `Month`

utils.go
package rotate

import (
	"fmt"
	"os"
	"path"
	"strings"
	"time"
)

// getNowRotateTime Used to obtain the naming format for time slices within a day.
// Preferably divisible by 24 Hour.
func nowRotateTime(dur time.Duration, loc *time.Location) time.Time {
	var base time.Time
	now := time.Now().In(loc)
	if loc != time.UTC {
		base = time.Date(now.Year(), now.Month(), now.Day(), now.Hour(), now.Minute(), now.Second(), now.Nanosecond(), time.UTC)
		base = base.Truncate(dur)
		base = time.Date(base.Year(), base.Month(), base.Day(), base.Hour(), base.Minute(), base.Second(), base.Nanosecond(), loc)
	} else {
		base = now.Truncate(dur)
	}
	return base
}

// getDirAllFiles get all files in a directory, when an error occurs panic.
func getDirAllFiles(dir string) []string {
	var result []string
	entries, err := os.ReadDir(dir)
	if err != nil {
		panic(err)
	}
	for _, entry := range entries {
		filename := join(dir, entry.Name())
		if entry.IsDir() {
			result = append(result, getDirAllFiles(filename)...)
		} else {
			result = append(result, filename)
		}
	}
	return result
}

// getEmptyDirs get all empty folders in a directory.
func getEmptyDirs(dir string) []string {
	var result []string
	dirs, err := os.ReadDir(dir)
	if err != nil {
		panic(err)
	}
	for _, entry := range dirs {
		if entry.IsDir() {
			name := join(dir, entry.Name())
			files, _ := os.ReadDir(name)
			if len(files) > 0 {
				entryDirs := getEmptyDirs(name)
				result = append(result, entryDirs...)
				var count int
				for _, file := range files {
					if file.IsDir() && hasString(entryDirs, join(name, file.Name())) {
						count++
					} else {
						break
					}
				}
				if count == len(files) {
					result = append(result, name)
				}
			} else {
				result = append(result, name)
			}
		}
	}
	return result
}

// hasString check if target exists in slice
func hasString(slice []string, target string) bool {
	for i := 0; i < len(slice); i++ {
		if slice[i] == target {
			return true
		}
	}
	return false
}

// join path, if it is not an absolute path add './'.
func join(elem ...string) string {
	p := path.Join(elem...)
	if len(elem) > 0 && strings.HasPrefix(elem[0], "./") {
		p = fmt.Sprintf("./%s", p)
	}
	return p
}

// calcDur calculate time interval
func calcDur(start, end time.Time) time.Duration {
	return end.Sub(start)
}


options.go
package rotate

import "time"

type Option func(*Rotate)

// WithRotationDuration set time interval, Preferably divisible by 24 Hour.
func WithRotationDuration(RotationDuration time.Duration) Option {
	return func(r *Rotate) {
		r.rotateDur = RotationDuration
	}
}

// WithTimeZone set time zone
func WithTimeZone(loc *time.Location) Option {
	return func(r *Rotate) {
		r.loc = loc
	}
}

// WithMaxAge max survival time
func WithMaxAge(MaxAge time.Duration) Option {
	return func(r *Rotate) {
		r.maxAge = MaxAge
	}
}

// WithMaxSize max file size
func WithMaxSize(MaxSize int64) Option {
	return func(r *Rotate) {
		r.maxSize = MaxSize
	}
}

// WithExpiredHandler expired file handler
func WithExpiredHandler(handler ExpFunc) Option {
	return func(r *Rotate) {
		r.expiredFunc = handler
	}
}

// WithDeleteEmptyFile whether to delete empty files
func WithDeleteEmptyFile(flag bool) Option {
	return func(r *Rotate) {
		r.deleteEmptyFileFlag = flag
	}
}

// WithDeleteEmptyDir whether to delete empty folders
func WithDeleteEmptyDir(flag bool) Option {
	return func(r *Rotate) {
		r.deleteEmptyDirFlag = flag
	}
}


rotate.go
package rotate

import (
	"fmt"
	"os"
	"path"
	"strconv"
	"strings"
	"sync"
	"time"
)

var (
	leftSymbol  byte        = '['
	rightSymbol byte        = ']'
	perm        os.FileMode = 0644
)

const (
	_  = iota
	KB = 1 << (10 * iota)
	MB = 1 << (10 * iota)
	GB = 1 << (10 * iota)
)

const (
	Day   = time.Hour * 24
	Week  = Day * 7
	Month = Day * 30
)

// SetSymbol Set the left and right symbols of the file slice sequence number.
func SetSymbol(left, right byte) {
	if left > 0 && right > 0 {
		leftSymbol = left
		rightSymbol = right
	}
}

// SetPerm Set permissions for creating files and folders.
func SetPerm(permissions os.FileMode) {
	perm = permissions
}

// ExpFunc Procedure for handler expired files.
type ExpFunc func([]string)

// Rotate The structure of the rotation file.
type Rotate struct {
	dir    string // storage folder path
	layout string // generate filename format template
	ext    string // file extension

	rotateDur           time.Duration  // rotation interval
	loc                 *time.Location // time zone settings
	maxAge              time.Duration  // maximum survival time
	maxSize             int64          // file size limit
	expiredFunc         ExpFunc        // processing of expired files
	deleteEmptyFileFlag bool           // remove empty file flag
	deleteEmptyDirFlag  bool           // remove empty dir flag

	mutex     *sync.RWMutex // lock
	index     int64         // rotated file index
	writeSize int64         // the size of the file content that has been written
	close     chan struct{} // close goroutine channel
	file      *os.File      // current file
	tick      *time.Timer   // timing rotation timer
}

// New create Rotate, Requires directory to store logs, time formatted layout, file extension.
func New(dir, layout, ext string, options ...Option) (*Rotate, error) {
	if len(dir) == 0 || len(layout) == 0 || len(ext) == 0 {
		return nil, fmt.Errorf("dir, layout, ext name len equel zero")
	}
	if !strings.HasPrefix(dir, "/") {
		if dir[1] != ':' && !strings.HasPrefix(dir, "./") {
			dir = fmt.Sprintf("./%s", dir)
		}
	}
	err := os.MkdirAll(dir, perm)
	if err != nil {
		return nil, err
	}
	if !strings.HasPrefix(ext, ".") {
		ext = fmt.Sprintf(".%s", ext)
	}
	r := &Rotate{
		// required parameters
		dir:    dir,
		layout: layout,
		ext:    ext,

		// optional parameters (defaults value)
		rotateDur:           Day,
		loc:                 time.Local,
		maxAge:              0,
		maxSize:             0,
		expiredFunc:         nil,
		deleteEmptyFileFlag: true,
		deleteEmptyDirFlag:  true,

		// private parameters
		mutex:     new(sync.RWMutex),
		index:     1,
		writeSize: 0,
		close:     make(chan struct{}),
	}
	for _, option := range options {
		option(r)
	}
	now := nowRotateTime(r.rotateDur, r.loc)
	ok, _, index := r.getNowFilenameMax(now)
	if ok {
		index++
	} else {
		index = 0
	}
	file, err := r.open(r.genFilename(now, index))
	if err != nil {
		return nil, err
	}
	r.file = file
	r.tick = time.NewTimer(calcDur(time.Now(), now.Add(r.rotateDur)))
	go r.task()
	r.expiredFileHandler()
	return r, nil
}

// implement the io.Writer interface
func (r *Rotate) Write(data []byte) (int, error) {
	var (
		n   int
		err error
	)
	r.mutex.RLock()
	if r.file == nil {
		err = fmt.Errorf("file is close")
	} else {
		n, err = r.file.Write(data)
	}
	r.mutex.RUnlock()

	if err != nil {
		return n, err
	}

	r.mutex.Lock()
	r.writeSize += int64(n)
	r.mutex.Unlock()

	if r.maxSize > 0 && r.writeSize >= r.maxSize {
		r.rotateSize()
	}
	return n, err
}

// Sync method
func (r *Rotate) Sync() error {
	return r.file.Sync()
}

// implement the io.Closer interface
func (r *Rotate) Close() error {
	r.close <- struct{}{}
	var f *os.File
	r.mutex.Lock()
	f = r.file
	r.file = nil
	r.mutex.Unlock()
	_ = f.Sync()
	return f.Close()
}

// goroutine tasks
func (r *Rotate) task() {
	for {
		select {
		case <-r.close:
			return
		case now := <-r.tick.C:
			r.tick.Reset(calcDur(now, nowRotateTime(r.rotateDur, r.loc).Add(r.rotateDur)))
			r.rotateTime(now)
			r.expiredFileHandler()
		}
	}
}

// rotateSize split by size
func (r *Rotate) rotateSize() {
	i := r.index + 1
	newFilename := r.genFilename(nowRotateTime(r.rotateDur, r.loc), i)
	f, err := r.open(newFilename)
	if err != nil {
		panic(err)
	}
	var oldFile *os.File

	r.mutex.Lock()
	oldFile = r.file
	r.file = f
	r.writeSize = 0
	r.index = i
	r.mutex.Unlock()

	if (i - 1) <= 1 {
		newFilename = r.genFilename(nowRotateTime(r.rotateDur, r.loc), 1)
		oldFilename := oldFile.Name()
		_ = oldFile.Sync()
		_ = oldFile.Close()
		_ = os.Rename(oldFilename, newFilename)
	} else {
		_ = oldFile.Sync()
		_ = oldFile.Close()
	}
}

// rotateTime split by time
func (r *Rotate) rotateTime(now time.Time) {
	filename := r.genFilename(now, 0)
	f, err := r.open(filename)
	if err != nil {
		panic(err)
	}

	var oldFile *os.File

	r.mutex.Lock()
	oldFile = r.file
	r.file = f
	r.writeSize = 0
	r.index = 1
	r.mutex.Unlock()

	if r.deleteEmptyFileFlag {
		filename = oldFile.Name()
		info, err := oldFile.Stat()
		_ = oldFile.Sync()
		_ = oldFile.Close()
		if err != nil {
			return
		}
		if info.Size() > 0 {
			return
		}
		_ = os.Remove(filename)
	} else {
		_ = oldFile.Sync()
		_ = oldFile.Close()
	}

	if r.deleteEmptyDirFlag {
		emptyDirs := getEmptyDirs(r.dir)
		for _, dir := range emptyDirs {
			_ = os.Remove(dir)
		}
	}
}

// open file by file name
func (r *Rotate) open(filename string) (*os.File, error) {
	dir, _ := path.Split(filename)
	err := os.MkdirAll(dir, perm)
	if err != nil {
		return nil, err
	}
	return os.OpenFile(filename, os.O_RDWR|os.O_CREATE|os.O_APPEND, perm)
}

// genFilename generate file name
func (r *Rotate) genFilename(t time.Time, i int64) string {
	var buf strings.Builder
	buf.WriteString(join(r.dir, t.Format(r.layout)))
	if i >= 1 {
		buf.WriteByte(leftSymbol)
		buf.WriteString(fmt.Sprintf("%d", i))
		buf.WriteByte(rightSymbol)
	}
	buf.WriteString(r.ext)
	return buf.String()
}

// parseFilename To parse the file name, the suffix must be removed
func (r *Rotate) parseFilename(filename string) (bool, time.Time, int64) {
	rightIndex := len(filename) - 1
	leftIndex := 0
	if filename[rightIndex] == rightSymbol {
		leftIndex = rightIndex
		for 0 <= leftIndex {
			if filename[leftIndex] == leftSymbol {
				break
			}
			leftIndex--
		}
		if leftIndex < 0 {
			return false, time.Time{}, 0
		}
		i, err := strconv.ParseInt(filename[leftIndex+1:rightIndex], 10, 64)
		if err != nil {
			return false, time.Time{}, 0
		}
		t, err := time.ParseInLocation(join(r.dir, r.layout), filename[:leftIndex], r.loc)
		if err != nil {
			return false, time.Time{}, 0
		}
		return true, t, i
	} else {
		t, err := time.ParseInLocation(join(r.dir, r.layout), filename, r.loc)
		if err != nil {
			return false, time.Time{}, 0
		}
		return true, t, 0
	}
}

// getNowFilenameMax get the max chunk at this time
func (r *Rotate) getNowFilenameMax(now time.Time) (bool, string, int64) {
	var result string
	var fileOk bool
	var maxIndex int64 = 0
	files := getDirAllFiles(r.dir)
	if len(files) <= 0 {
		return fileOk, result, 0
	}
	for i := 0; i < len(files); i++ {
		ok, t, index := r.parseFilename(strings.TrimSuffix(files[i], r.ext))
		if ok && now.Equal(t) && index >= maxIndex {
			fileOk = true
			maxIndex = index
			result = files[i]
		}
	}
	return fileOk, result, maxIndex
}

// handle expired files
func (r *Rotate) expiredFileHandler() {
	if r.expiredFunc != nil && r.maxAge > 0 {
		exp := nowRotateTime(r.rotateDur, r.loc).Add(-r.maxAge)
		result := r.getExpirationFiles(exp)
		// prevent panic
		defer func() {
			if err := recover(); err != nil {
				fmt.Printf("expired handler panic:%v", err)
			}
		}()
		r.expiredFunc(result)
	}
}

// get expired files
func (r *Rotate) getExpirationFiles(exp time.Time) []string {
	var result []string
	files := getDirAllFiles(r.dir)
	for i := 0; i < len(files); i++ {
		ok, t, _ := r.parseFilename(strings.TrimSuffix(files[i], r.ext))
		if ok && (exp.Equal(t) || exp.After(t)) {
			result = append(result, files[i])
		}
	}
	return result
}
